File: ftl/samples/java/TibRecvEx.java
/*
 * Copyright (c) 2010-2018 TIBCO Software Inc.
 * Licensed under a BSD-style license. Refer to [LICENSE]
 * For more information, please contact:
 * TIBCO Software Inc., Palo Alto, California, USA
 */

/*
 * This is a sample of a basic java FTL subscriber program which receives the
 * requested number of messages and then cleans up and exits.
 */

package com.tibco.ftl.samples;

import java.util.*;

import com.tibco.ftl.*;

public class TibRecvEx implements SubscriberListener
{
    int      count          = 1;
    int      recved         = 0;
    String   appName        = "tibrecv";
    String   realmServer    = null;
    String   defaultUrl     = "http://localhost:8080";
    String   defaultSecureUrl = "https://localhost:8080";
    String   match          = null;
    String   subEndpoint    = "tibrecv-endpoint";
    String   trcLevel       = null;
    String   user           = "guest";
    String   password       = "guest-pw";
    String   identifier     = null;
    String   durableName    = null;
    String   secondary      = null;
    boolean  explicitAck    = false;
    boolean  trustAll       = false;
    String   trustFile      = null;
    boolean  unsubscribe    = false;
    String   clientLabel    = null;

    // MessageFiledRef objects could also be used instead for faster access
    final String   myLong         = "My-Long";
    final String   myString       = "My-String";
    final String   myOpaque       = "My-Opaque";

    List<Message>   ackableMsgs = new LinkedList<Message>();
    
    String   usageString    =
        "TibRecvEx [options] url\n" +
        " Default url is " + defaultUrl + "\n" +
        "  where options can be:\n" +
        "    -a, --application <name>\n" +
        "    -cl, --client-label <string>\n" +
        "    -c, --count <int>\n" +
        "    -d, --durablename <name>\n" +
        "    -e, --endpoint <name>\n" +
        "    -h, --help\n" +
        "    -id,--identifier <string>\n" +
        "    -m, --match <content match>\n" +
        "           where <content match> is of the form:\n" +
        "                 '{\"fieldname\":<value>}'\n" +
        "           where <value> is:\n" +
        "                 \"string\" for a string field\n" +
        "                 a number such as 2147483647 for an integer field\n" +
        "    -p, --password <string>\n" +
        "    -s, --secondary <string>\n" +
        "    -t, --trace <level>\n" +
        "           where <level> can be:\n" +
        "                 off, severe, warn, info, debug, verbose\n" +
        "        --trustall\n" +
        "        --trustfile <file>\n" +
        "    -us, --unsubscribe\n" +
        "    -u, --user <string>\n" +
        "    -x, --explicitAck \n";

    public TibRecvEx(String[] args)
    {
        System.out.printf("#\n# %s\n#\n# %s\n#\n",
                          this.getClass().getName(),
                          FTL.getVersionInformation());

        parseArgs(args);
    }
    
    public void usage()
    {
        System.out.printf(usageString);
        System.exit(0);
    }

    public void parseArgs(String args[])
    {
        int i           = 0;
        int argc        = args.length;
        String s        = null;
        int n           = 0;
        TibAux aux      = new TibAux(args, usageString);

        for (i = 0; i < argc; i++)
        {
            s = aux.getString(i, "--application", "-a");
            if (s != null) {
                appName = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--durablename", "-d");
            if (s != null) {
                durableName = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--client-label", "-cl");
            if (s != null) {
                clientLabel = s;
                i++;
                continue;
            }
            n = aux.getInt(i, "--count", "-c");
            if (n > -1) {
                count = n;
                i++;
                continue;
            }
            s = aux.getString(i, "--endpoint", "-e");
            if (s != null) {
                subEndpoint = s;
                i++;
                continue;
            }
            if (aux.getFlag(i, "--help", "-h")) {
                usage();
            }
            s = aux.getString(i, "--identifier", "-id");
            if (s != null) {
                identifier = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--match", "-m");
            if (s != null) {
                match = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--password", "-p");
            if (s != null) {
                password = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--secondary", "-s");
            if (s != null) {
                secondary = s;
                i++;
                continue;
            }
            s = aux.getString(i, "--trace", "-t");
            if (s != null) {
                trcLevel = s;
                i++;
                continue;
            }
            if (aux.getFlag(i, "--trustall", "--trustall")) {
                if (trustFile != null)
                {
                    System.out.println("cannot specify both --trustall and --trustfile");
                    usage();
                }
                trustAll = true;
                continue;
            }
            s = aux.getString(i, "--trustfile", "--trustfile");
            if (s != null)
            {
                trustFile = s;
                if (trustAll)
                {
                    System.out.println("cannot specify both --trustall and --trustfile");
                    usage();
                }
                i++;
                continue;
            }
            s = aux.getString(i, "--user", "-u");
            if (s != null) {
                user = s;
                i++;
                continue;
            }
            if (aux.getFlag(i, "--explicitAck", "-x")) {
                explicitAck = true;
                continue;
            }
            if (aux.getFlag(i, "--unsubscribe", "-us")) {
                unsubscribe = true;
                continue;
            }
            if (args[i].startsWith("-", 0)) {
                System.out.println("invalid option: "+args[i]);
                usage();
            }
            realmServer = args[i];
        }

        System.out.print("Invoked as: TibRecvEx ");
        for (i = 0; i < argc; i++)
            System.out.print(args[i]+" ");
        System.out.println("");
    }

    private void printOpaque(byte []opaque)
    {
        String outputString;

        System.out.print(" type opaque:  size " + opaque.length);

        String s = new String(opaque);
        
        if (s.length() > 60)
            outputString = s.substring(0, 60) + "<truncated>";
        else
            outputString = s;
        
         System.out.print(" data: " + outputString);
         if (s != outputString)
             System.out.print("<truncated>");
         System.out.println("\n");
    }
    
    public void printMessage(
        Message         msg) throws FTLException
    {
        int fieldCount = 0;
        System.out.println("message:");
        
        if (msg.isFieldSet(myLong))
        {
            System.out.println(" type long: " + msg.getLong(myLong));
            fieldCount++;
        }
        if(msg.isFieldSet(myString))
        {
            System.out.println(" type string: " + msg.getString(myString));
            fieldCount++;
        }
        if (msg.isFieldSet(myOpaque))
        {
            byte[] opaque = msg.getOpaque(myOpaque);
            printOpaque(opaque);
            fieldCount++;
        }
        
        if (fieldCount != 3)
        {
            System.out.println(" " + msg);
        }        
    }

    public void messagesReceived(List<Message> messages, EventQueue eventQueue)
    {
        int i;
        int msgNum = messages.size();
         
        for (i = 0;  i < msgNum;  i++) 
        {
            try
            {
                //
                // Do not destroy message as it is owned by the library.  If
                // you need to pass the messages to something outside this
                // callback, use Message.mutableCopy() and pass the copies.
                // 
                recved++;
                System.out.printf("received message %d \n", recved);
                printMessage(messages.get(i));               

                if (explicitAck)
                    ackableMsgs.add(messages.get(i).mutableCopy());
            }
            catch(FTLException exp)
            {
                exp.printStackTrace();
            }
        }
    }
    
    public int recv() throws FTLException
    {
        TibProperties   props;
        Realm           realm;
        ContentMatcher  cm = null;
        ContentMatcher  am;
        Subscriber      sub;
        Subscriber      adv;
        EventQueue      queue;
        
        // Set global trace to specified level
        if (trcLevel != null)
            FTL.setLogLevel(trcLevel);
        
        // Get a single realm per process
        props = FTL.createProperties();
        props.set(Realm.PROPERTY_STRING_USERNAME, user);
        props.set(Realm.PROPERTY_STRING_USERPASSWORD, password);
        if (identifier != null)
            props.set(Realm.PROPERTY_STRING_APPINSTANCE_IDENTIFIER, identifier);
        if (secondary != null)
            props.set(Realm.PROPERTY_STRING_SECONDARY_SERVER, secondary);

        if (trustAll || trustFile != null)
        {
            if (realmServer == null)
                realmServer = defaultSecureUrl;

            if (trustAll)
            {
                props.set(Realm.PROPERTY_LONG_TRUST_TYPE, Realm.HTTPS_CONNECTION_TRUST_EVERYONE);
            }
            else
            {
                props.set(Realm.PROPERTY_LONG_TRUST_TYPE, Realm.HTTPS_CONNECTION_USE_SPECIFIED_TRUST_FILE);
                props.set(Realm.PROPERTY_STRING_TRUST_FILE, trustFile);
            }
        }
        else
        {
            if (realmServer == null)
                realmServer = defaultUrl;
        }

        if (clientLabel != null)
            props.set(Realm.PROPERTY_STRING_CLIENT_LABEL, clientLabel);
        else
            props.set(Realm.PROPERTY_STRING_CLIENT_LABEL, this.getClass().getSimpleName());

        realm = FTL.connectToRealmServer(realmServer, appName, props);
        props.destroy();
        
        // Create a content matcher, if specified
        if (match != null)
            cm = realm.createContentMatcher(match);

        props = null;
        props = FTL.createProperties();
        
        if (explicitAck)
            props.set(Subscriber.PROPERTY_BOOL_EXPLICIT_ACK, explicitAck);

        if (durableName != null)
            props.set(Subscriber.PROPERTY_STRING_DURABLE_NAME, durableName);

        // Create a subscriber
        sub = realm.createSubscriber(subEndpoint, cm, props);

        // Create a data loss advisory matcher
        am = realm.createContentMatcher("{\""+Advisory.FIELD_NAME+
                                        "\":\""+Advisory.NAME_DATALOSS+"\"}");

        // Create an advisory subscriber
        adv = realm.createSubscriber(Advisory.ENDPOINT_NAME, am);
        am.destroy();

        // Create a queue and add subscriber to it
        props = FTL.createProperties();
        queue = realm.createEventQueue(props);
        props.set(EventQueue.PROPERTY_INT_DISCARD_POLICY, EventQueue.DISCARD_NEW);
        props.set(EventQueue.PROPERTY_INT_DISCARD_POLICY_MAX_EVENTS, 1000);
        queue.addSubscriber(sub, this);
        queue.addSubscriber(adv, new AdvisoryListener());
        props.destroy();
        props = null;
        
        // Begin dispatching messages
        System.out.printf("waiting for message(s)\n");
        while (recved < count)
        {
            queue.dispatch();
            while (ackableMsgs.size() > 0)
            {
                Message msg = ackableMsgs.remove(0);
                msg.acknowledge();
                msg.destroy();
            }
        }

        // Clean up
        queue.removeSubscriber(sub);
        queue.removeSubscriber(adv);
        queue.destroy();
        sub.close();
        adv.close();
        if (cm != null)
            cm.destroy();
        if (unsubscribe && durableName != null)
            realm.unsubscribe(subEndpoint, durableName);
        realm.close();

        return 0;
    }
    
    public static void main(String[] args)
    {
        TibRecvEx s  = new TibRecvEx(args);
        try
        {
            s.recv();
        }
        catch (Throwable e)
        {
            e.printStackTrace();
        }
    } 

    private class AdvisoryListener implements SubscriberListener
    {
        public void messagesReceived(List<Message> messages, EventQueue eventQueue)
        {
            int msgNum = messages.size();

            for (int i = 0;  i < msgNum;  i++)
            {
                Message msg = messages.get(i);
                try
                {
                    System.out.println("advisory:");

                    System.out.println("Name: " + msg.getString(Advisory.FIELD_NAME));

                    System.out.println("    Severity: " +
                                       msg.getString(Advisory.FIELD_SEVERITY));
                    System.out.println("    Module: " +
                                       msg.getString(Advisory.FIELD_MODULE));
                    System.out.println("    Reason: " +
                                       msg.getString(Advisory.FIELD_REASON));
                    System.out.println("    Timestamp: " +
                                       msg.getDateTime(Advisory.FIELD_TIMESTAMP));
                    System.out.println("    Aggregation Count: " +
                                       msg.getLong(Advisory.FIELD_AGGREGATION_COUNT));
                    System.out.println("    Aggregation Time: " +
                                       msg.getDouble(Advisory.FIELD_AGGREGATION_TIME));

                    if (msg.isFieldSet(Advisory.FIELD_ENDPOINTS))
                    {
                        System.out.println("    Endpoints:");
                        String[] sarray = msg.getStringArray(Advisory.FIELD_ENDPOINTS);
                        for (int j = 0; j < sarray.length; j++)
                            System.out.println("      " + sarray[j]);
                    }

                    if (msg.isFieldSet(Advisory.FIELD_SUBSCRIBER_NAME))
                    {
                        System.out.println("  Subscriber Name: " +
                                           msg.getString(Advisory.FIELD_SUBSCRIBER_NAME));
                    }
                }
                catch(FTLException exp)
                {
                    exp.printStackTrace();
                }
            }
        }
    }
}

File: ftl/samples/c/tibrecvex.c
/*
 * Copyright (c) 2010-2018 TIBCO Software Inc.
 * Licensed under a BSD-style license. Refer to [LICENSE]
 * For more information, please contact:
 * TIBCO Software Inc., Palo Alto, California, USA
 */

/*
 * This is a sample of a basic C FTL subscriber program which receives the
 * requested number of messages and then cleans up and exits.
 */

#include <stdlib.h>
#include <stdio.h>
#include <stddef.h>
#include <string.h>
#include <ctype.h>

#include "tib/ftl.h"

#include "tibaux.h"

#define DEFAULT_URL "http://localhost:8080"
#define DEFAULT_SECURE_URL "https://localhost:8080"

tibint32_t      count           = 1;
tibint32_t      recved          = 0;
const char      *appName        = "tibrecv";
const char      *realmServer    = NULL;
const char      *match          = NULL;
const char      *subEndpoint    = "tibrecv-endpoint";
const char      *trcLevel       = NULL;
const char      *user           = "guest";
const char      *password       = "guest-pw";
const char      *identifier     = NULL;
const char      *secondary      = NULL;
const char      *durableName    = NULL;
tibbool_t       explicitAck     = tibfalse;
tibbool_t       unsubscribe     = tibfalse;
tibbool_t       trustAll        = tibfalse;
const char      *trustFile      = NULL;
const char      *clientLabel    = NULL;

tibFieldRef     refLong         = NULL;
tibFieldRef     refString       = NULL;
tibFieldRef     refOpaque       = NULL;

typedef struct {
    tibint32_t      n;
    tibMessage      msgs[250];
} ackListRec, *ackList;

void
usage(
    int         argc,
    char        **argv)
{
    printf("Usage: %s [options] url\n", argv[0]);
    printf("Default url is "DEFAULT_URL"\n");
    printf("    -a,  --application <name>\n"
           "    -d,  --durablename <name>\n"
           "    -c,  --count <int>\n"
           "         --client-label <label>\n"
           "    -e,  --endpoint <name>\n"
           "    -h,  --help\n"
           "    -id, --identifier <name>\n"
           "    -m,  --match <content match>\n"
           "            where <content match> is of the form:\n"
           "                 '{\"fieldname\":<value>}'\n"
           "            where <value> is:\n"
           "                 \"string\" for a string field\n"
           "                 a number such as 2147483647 for an integer field\n"
           "    -p,  --password <string>\n"
           "    -u,  --user <string>\n"
           "    -s,  --secondary <string>\n"
           "    -t,  --trace <level>\n"
           "            where <level> can be:\n"
           "                  off, severe, warn, info, debug, verbose\n"
           "         --trustall\n"
           "         --trustfile <file>\n"
           "    -us  --unsubscribe \n"
           "            the durable name needs to be specified as well\n"
           "    -x,  --explicitAck \n");

    exit(1);
}

void
parseArgs(int argc, char** argv)
{
    int    i = 1;

    tibAux_InitGetArgs(argc, argv, usage);

    for (i = 1;  i < argc;  i++)
    {
        if (tibAux_GetString(&i, "--application", "-a", &appName))
            continue;
        if (tibAux_GetString(&i, "--client-label", "--client-label", &clientLabel))
            continue;
        if (tibAux_GetString(&i, "--durablename", "-d", &durableName))
            continue;
        if (tibAux_GetInt(&i ,"--count", "-c", &count))
            continue;
        if (tibAux_GetString(&i, "--endpoint", "-e", &subEndpoint))
            continue;
        if (tibAux_GetFlag(&i, "--help", "-h"))
            usage(argc, argv);
        if (tibAux_GetString(&i, "--identifier", "-id", &identifier))
            continue;
        if (tibAux_GetString(&i, "--match", "-m", &match))
            continue;
        if (tibAux_GetString(&i, "--password", "-p", &password))
            continue;
        if (tibAux_GetString(&i, "--secondary", "-s", &secondary))
            continue;
        if (tibAux_GetString(&i, "--trace", "-t", &trcLevel))
            continue;
        if (tibAux_GetString(&i, "--user", "-u", &user))
            continue;
        if (tibAux_GetFlag(&i, "--explicitAck", "-x"))
        {
            explicitAck = tibtrue;
            continue;
        }
        if (tibAux_GetFlag(&i, "--unsubscribe", "-us"))
        {
            unsubscribe = tibtrue;
            continue;
        }
        if (tibAux_GetFlag(&i, "--trustall", "--trustall")) {
            if (trustFile)
            {
                printf("cannot specify both --trustall and --trustfile\n");
                usage(argc, argv);
            }
            trustAll = tibtrue;
            continue;
        }
        if (tibAux_GetString(&i, "--trustfile", "--trustfile", &trustFile))
        {
            if (trustAll)
            {
                printf("cannot specify both --trustall and --trustfile\n");
                usage(argc, argv);
            }
            continue;
        }
        if (strncmp(argv[i], "-", 1)==0)
        {
            printf("invalid option: %s\n", argv[i]);
            usage(argc, argv);
        }
        realmServer = argv[i];
    }

    printf("Invoked as: ");
    for (i = 0;  i < argc;  i++)
        printf("%s ", argv[i]);
    printf("\n");
}

void
memFail(
    const char          *obj)
{
    printf("memory allocation failed for %s\n", obj);
    exit(-1);
}

void
createMsgFieldRefs(
    tibEx               ex)
{
    refLong         = tibFieldRef_Create(ex, "My-Long");
    refString       = tibFieldRef_Create(ex, "My-String");
    refOpaque       = tibFieldRef_Create(ex, "My-Opaque");
}

void
cleanupMsgFieldRefs(
    tibEx               ex)
{
    tibFieldRef_Destroy(ex, refLong);
    tibFieldRef_Destroy(ex, refString);
    tibFieldRef_Destroy(ex, refOpaque);
}

void
printOpaque(
    tibEx               ex,
    const void          *buffer,
    tibint32_t          size)
{
    const char          *val = (const char*)buffer;
    tibint32_t          i;
    tibint32_t          unprintable = 0;

    if (tibEx_GetErrorCode(ex) != TIB_OK)
        return;

    printf("  type opaque: size %d data: '", size);
    for (i = 0; i < size; i++)
        if (isprint(val[i]))
            printf("%c", val[i]);
        else
            unprintable++;
    printf("'");
    if (unprintable > 0)
        printf(" (unprintable chars %d)", unprintable);
    printf("\n");
}

void
printMessage(
    tibEx               ex,
    tibMessage          msg)
{
    tibint64_t          lvalue = 0;
    const char          *svalue = NULL;
    const void          *ovalue = NULL;
    tibint32_t          size = 0;
    tibint32_t          len = 0;
    char                *buffer = 0;
    tibint32_t          fieldCount = 0;

    printf("message:\n");

    // we assume this has the fields set by tibsend.c, if not just
    // print the message and be done with it.
    if (tibMessage_IsFieldSetByRef(ex, msg, refLong))
    {
        lvalue = tibMessage_GetLongByRef(ex, msg, refLong);
        printf("  type long:  %"auxPRId64"\n", lvalue);
        fieldCount++;
    }
    if (tibMessage_IsFieldSetByRef(ex, msg, refString))
    {
        svalue = tibMessage_GetStringByRef(ex, msg, refString);
        printf("  type string: '%s'\n", svalue ? svalue : "<null>");
        fieldCount++;
    }
    if (tibMessage_IsFieldSetByRef(ex, msg, refOpaque))
    {
        ovalue = tibMessage_GetOpaqueByRef(ex, msg, refOpaque, &size);
        printOpaque(ex, ovalue, size);
        fieldCount++;
    }

    if (fieldCount != 3)
    {
        // get the size of output string
        len = tibMessage_ToString(ex, msg, NULL, 0);
        buffer = (char*)malloc(len);
        if (buffer)
        {
            (void)tibMessage_ToString(ex, msg, buffer, len);
            printf("  %s\n", buffer);
            free(buffer);
        }
    }

    if (tibEx_GetErrorCode(ex) != TIB_OK)
        tibAux_PrintException(ex);
    fflush(stdout);
}

void
onMessages(
    tibEx               ex,
    tibEventQueue       msgQueue,
    tibint32_t          msgNum,
    tibMessage          *msgs,
    void                **closures)
{
    tibint32_t          i;

    for (i = 0; i < msgNum; i++)
    {
        //
        // Do not destroy message as it is owned by the library.  If you need
        // to pass the messages to something outside this callback, use
        // tibMessage_MutableCopy and pass the copies.
        //
        recved++;
        printf("received message %d\n", recved);
        printMessage(ex, msgs[i]);

        if (explicitAck)
        {
            ackList     waitingAcks = (ackList) closures[i];

            waitingAcks->msgs[waitingAcks->n++] = tibMessage_MutableCopy(ex, msgs[i]);
        }
    }

    // any exception that happens in the callback does not
    // propagate to the caller of the tibEventQueue_Dispatch
    if (tibEx_GetErrorCode(ex) != TIB_OK)
        tibAux_PrintException(ex);
}

void
onAdvisory(
    tibEx               ex,
    tibEventQueue       msgQueue,
    tibint32_t          msgNum,
    tibMessage          *msgs,
    void                **closures)
{
    tibint32_t          i;

    for (i = 0; i < msgNum; i++)
    {
        printf("received advisory\n");
        tibAux_PrintAdvisory(ex, msgs[i]);
    }

    // any exception that happens in the callback does not
    // propagate to the caller of the tibEventQueue_Dispatch
    if (tibEx_GetErrorCode(ex) != TIB_OK)
        tibAux_PrintException(ex);
}

int
main(int argc, char** argv)
{
    tibEx               ex;
    tibProperties       props;
    tibRealm            realm;
    tibEventQueue       queue = NULL;
    tibContentMatcher   cm = NULL;
    tibContentMatcher   am;
    tibSubscriber       sub;
    tibSubscriber       adv;
    char                str[256];
    ackListRec          ackableMsgs;

    printf("#\n# %s\n#\n# %s\n#\n", argv[0], tib_Version());

    parseArgs(argc, argv);

    ex = tibEx_Create();
    if (ex == NULL)
        memFail("ex");

    tib_Open(ex, TIB_COMPATIBILITY_VERSION);
    CHECK(ex);

    // Set global trace to specified level
    if (trcLevel != NULL)
        tib_SetLogLevel(ex, trcLevel);

    // Get a single realm per process
    props = tibProperties_Create(ex);
    tibProperties_SetString(ex, props,
                            TIB_REALM_PROPERTY_STRING_USERNAME, user);
    tibProperties_SetString(ex, props,
                            TIB_REALM_PROPERTY_STRING_USERPASSWORD, password);
    if (identifier)
        tibProperties_SetString(ex, props,
                                TIB_REALM_PROPERTY_STRING_APPINSTANCE_IDENTIFIER, identifier);
    if (secondary)
        tibProperties_SetString(ex, props,
                                TIB_REALM_PROPERTY_STRING_SECONDARY_SERVER, secondary);

    if (trustAll || trustFile)
    {
        if (!realmServer)
            realmServer = DEFAULT_SECURE_URL;

        if (trustAll)
        {
            tibProperties_SetLong(ex, props, TIB_REALM_PROPERTY_LONG_TRUST_TYPE,
                                        TIB_REALM_HTTPS_CONNECTION_TRUST_EVERYONE);
        }
        else
        {
            tibProperties_SetLong(ex, props, TIB_REALM_PROPERTY_LONG_TRUST_TYPE,
                    TIB_REALM_HTTPS_CONNECTION_USE_SPECIFIED_TRUST_FILE);
            tibProperties_SetString(ex, props, TIB_REALM_PROPERTY_STRING_TRUST_FILE, trustFile);
        }
    }
    else
    {
        if (!realmServer)
            realmServer = DEFAULT_URL;
    }

    // set the client label
    if (clientLabel != NULL)
        tibProperties_SetString(ex, props, TIB_REALM_PROPERTY_STRING_CLIENT_LABEL, clientLabel);
    else
        tibProperties_SetString(ex, props, TIB_REALM_PROPERTY_STRING_CLIENT_LABEL, argv[0]);

    realm = tibRealm_Connect(ex, realmServer, appName, props);
    CHECK(ex);
    tibProperties_Destroy(ex, props);

    // Create a content matcher, if specified
    if (match != NULL)
        cm = tibContentMatcher_Create(ex, realm, match);

    // Create a subscriber
    props = tibProperties_Create(ex);

    if (explicitAck)
        tibProperties_SetBoolean(ex, props,
                                 TIB_SUBSCRIBER_PROPERTY_BOOL_EXPLICIT_ACK, explicitAck);

    if (durableName)
        tibProperties_SetString(ex, props,
                                TIB_SUBSCRIBER_PROPERTY_STRING_DURABLE_NAME, durableName);

    sub = tibSubscriber_Create(ex, realm, subEndpoint, cm, props);
    CHECK(ex);
    tibProperties_Destroy(ex, props);

    // Create a data loss advisory matcher 
    tibAux_snprintf(str, sizeof(str), "{\"%s\":\"%s\"}", 
                  TIB_ADVISORY_FIELD_NAME, TIB_ADVISORY_NAME_DATALOSS);
    am = tibContentMatcher_Create(ex, realm, str);

    // Create an advisory subscriber
    adv = tibSubscriber_Create(ex, realm, TIB_ADVISORY_ENDPOINT_NAME, am, NULL);
    CHECK(ex);

    // Create a queue and add subscribers to it
    props = tibProperties_Create(ex);
    tibProperties_SetInt(ex, props, TIB_EVENTQUEUE_PROPERTY_INT_DISCARD_POLICY, TIB_EVENTQUEUE_DISCARD_NEW);
    tibProperties_SetInt(ex, props, TIB_EVENTQUEUE_PROPERTY_INT_DISCARD_POLICY_MAX_EVENTS, 1000);
    queue = tibEventQueue_Create(ex, realm, props);
    tibProperties_Destroy(ex, props);
    tibEventQueue_AddSubscriber(ex, queue, sub, onMessages, &ackableMsgs);
    tibEventQueue_AddSubscriber(ex, queue, adv, onAdvisory, NULL);
    CHECK(ex);

    // set up references to the expected message format for fast access
    createMsgFieldRefs(ex);

    // Begin dispatching messages
    printf("waiting for message(s)\n");
    while (recved < count)
    {
        tibint32_t          i;

        ackableMsgs.n = 0;
        tibEventQueue_Dispatch(ex, queue, TIB_TIMEOUT_WAIT_FOREVER);
        if (tibEx_GetErrorCode(ex) != TIB_OK)
            break;

        for (i = 0;  i < ackableMsgs.n;  i++)
        {
            tibMessage_Acknowledge(ex, ackableMsgs.msgs[i]);
            if (tibEx_GetErrorCode(ex) != TIB_OK)
            {
                tibAux_PrintException(ex);
                tibEx_Clear(ex);
            }

            tibMessage_Destroy(ex, ackableMsgs.msgs[i]);
        }
    }

    // Clean up
    cleanupMsgFieldRefs(ex);
    if (queue && sub)
        tibEventQueue_RemoveSubscriber(ex, queue, sub, NULL);
    if (queue && adv)
        tibEventQueue_RemoveSubscriber(ex, queue, adv, NULL);
    tibEventQueue_Destroy(ex, queue, NULL);
    tibSubscriber_Close(ex, sub);
    tibSubscriber_Close(ex, adv);
    tibContentMatcher_Destroy(ex, cm);
    tibContentMatcher_Destroy(ex, am);

    if (durableName && unsubscribe)
    {
        tibRealm_Unsubscribe(ex, realm, subEndpoint, durableName);
    }

    tibRealm_Close(ex, realm);
    tib_Close(ex);

    // Check if it worked before destroying the exception
    CHECK(ex);
    tibEx_Destroy(ex);

    return 0;
}
